aesd-char-driver/main.c

```diff
@@ -0,0 +1,268 @@
+/**
+ * @file aesdchar.c
+ * @brief Functions and data related to the AESD char driver implementation
+ *
+ * Based on the implementation of the "scull" device driver, found in
+ * Linux Device Drivers example code.
+ *
+ * @author Dan Walkes
+ * @date 2019-10-22
+ * @copyright Copyright (c) 2019
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/printk.h>
+#include <linux/types.h>
+#include <linux/slab.h>
+#include <linux/cdev.h>
+#include <linux/fs.h> // file_operations
+
+#include "aesd-circular-buffer.h"
+#include "aesdchar.h"
+int aesd_major = 0; // use dynamic major
+int aesd_minor = 0;
+
+MODULE_AUTHOR("Chinmay Shalawadi"); /** TODO: fill in your name **/
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    struct aesd_dev *dev;
+
+    PDEBUG("open");
+
+    dev = container_of(inode->i_cdev, struct aesd_dev, cdev);
+    filp->private_data = dev;
+
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+    /**
+     * TODO: handle release
+     */
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                  loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    size_t entry_offset;
+    int buff_count = 0;
+    struct aesd_dev *dev_strcut = (struct aesd_dev *)filp->private_data;
+    struct aesd_buffer_entry *entry;
+    PDEBUG("read %zu bytes with offset %lld", count, *f_pos);
+
+    mutex_lock(&(dev_strcut->lock));
```

Comment:
@neohere97 consider using the interruptible versions here so you can exit, and check the return codes/handle errors.